### PR TITLE
feat: multipart copy

### PR DESCRIPTION
**Pull request by oliverdaff**
## 📢 What

What changes have been made within this PR?

This PR introduces a new `S3MultipartCopier` struct for handling multipart copy operations in S3. The changes include:
- Implementing the `S3MultipartCopier` struct to efficiently copy large files (greater than 5GB) using multipart copy.
- Providing custom error handling for multipart copy operations.
- Integration tests to verify the functionality of multipart copy.

## ❓ Why

Why are we submitting this PR?

Currently, S3 copy operations only work for files up to 5GB. Multipart copy is not only necessary for larger files but is also much faster for files under 5GB due to parallel copies. However, multipart copy will make more API requests, potentially increasing costs. This PR addresses the need for efficient, scalable, and reliable copying of large objects in S3.

## 🚦 Depends on

Are there any other PRs that need to be merged first?

* No, this PR is self-contained and does not depend on any other PRs.

## 😟 Concerns

* No major concerns at this time. However, it's important to note the potential increase in costs due to the higher number of API requests made by multipart copy operations.

## 📝 Notes

* See the [data engineering handbook](https://dataengineering.harrisonai.io/pages/source_control.html#anatomy-of-a-pull-request) for more info.

### CODE REVIEW

File: src/s3/multipartcopy.rs [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+    /// # Returns
+    ///
+    /// An empty result indicating success.
+    #[instrument(skip(self))]
+    pub async fn send(&self) -> Result<(), S3MultipartCopierError> {
+        tracing::info!("Starting multipart copy");
+        let source_size =
+            get_source_size(&self.client, self.source.bucket(), self.source.key()).await?;
+        let part_count =
+            ((f64::value_from(source_size)? / f64::value_from(self.part_size)?).ceil()).approx()?;
```
Comment (timleslie) [inline src/s3/multipartcopy.rs:205]: **suggestion:** This piece of code is begging for an edge case bug. Would be good to have explicit tests for either side of the edges to make 100% sure it does the right thing. Either as additional test cases of the full function, or factored out into a helper method with individual tests.

File: src/s3/multipartcopy.rs [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+#[derive(Debug, TypedBuilder)]
+struct CopyUploadPart<'a> {
+    src: &'a S3Object,
+    dst: &'a S3Object,
+    upload_id: &'a str,
+    part_number: i32,
+    byte_range: (i64, i64),
+}
+
+/// A struct to handle S3 multipart copy operations.
```
Comment (timleslie) [inline src/s3/multipartcopy.rs:78]: **suggestion:** Could we add some more docs on how to use this feature? It looks like we get some docs for free from `TypedBuilder`, but they're not great. 

Screenshot 2024-06-03 at 11 07 19 AM

In particular, there's no documentation of the individual fields, their types, or how they should be used in the builder. I guess we could work around this by making the fields `pub` and adding docstrings, or else perhaps an example, like below (but formatted correctly 😅).

In particular, it would be helpful to have some guidance on the `part_size` and `max_concurrent_uploads` fields, and when/why the user might want to set them.

```suggestion
/// A struct to handle S3 multipart copy operations.
///
///        let dest = S3Object::new(test_bucket, dst_key);
///        let copyier = S3MultipartCopier::builder()
///            .client(client.clone())
///            .source(src)
///            .destination(dest.clone())
///            .part_size((5 * MIB).try_into()?)
///            .max_concurrent_uploads(2)
///            .build();
///        copyier.send().await?;
```

You can run the following command to generate the docs locally
```sh
RUSTDOCFLAGS="--cfg docsrs" cargo +nightly doc --open --all-features
```

File: src/s3/multipartcopy.rs [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+    client: &Client,
+    bucket: &str,
+    key: &str,
+) -> Result<i64, aws_sdk_s3::error::SdkError<HeadObjectError>> {
+    let head_object = client.head_object().bucket(bucket).key(key).send().await?;
+
+    Ok(head_object.content_length().unwrap_or(0))
+}
+
+const DEFAULT_COPY_PART_SIZE: i64 = 50 * MIB as i64;
```
Comment (timleslie) [inline src/s3/multipartcopy.rs:43]: **question;** Is the choice of 50MB here significant? Could we leave a small comment justifying this choice. If it's mostly arbitrary then we can just mention that 👍

**Review by timleslie — CHANGES_REQUESTED** [review]
This looks great. The biggest thing missing I think is an example in the published docs showing how to use it. Other than that, the other suggestions are mostly cosmetic.

File: src/s3/multipartcopy.rs (hunk 2 above)
Comment (timleslie) [reply src/s3/multipartcopy.rs:78]: > However, it's important to note the potential increase in costs due to the higher number of API requests made by multipart copy operations.

Would be useful to leave a comment for the user about this in this documentation block 👍

**Comment by timleslie** [comment]
Could you also update the `CHANGELOG.md` with an entry under the next release (`0.13.4`) to make life easier for whoever makes the next release 🙏

File: src/s3/s3_object.rs [hunk 4]
```diff
@@ -26,6 +26,14 @@ impl S3Object {
...
         }
     }
+
+    pub fn bucket(&self) -> &str {
+        &self.bucket
+    }
+
+    pub fn key(&self) -> &str {
+        &self.key
+    }
```
Comment (timleslie) [inline src/s3/s3_object.rs:36]: **question:** What's the motivation for the additional methods here? My preference would be to keep the API as small as possible and have the user access the fields directly for this relatively simple struct.

```suggestion
```

File: src/s3/multipartcopy.rs [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+                        .build();
+                    tracing::debug!(part= ?part ,"Copying");
+                    self.upload_part_copy(part).await
+                }
+            })
+            .collect();
+
+        let completed_parts: Vec<CompletedPart> = stream::iter(parts)
+            .buffer_unordered(self.max_concurrent_uploads)
+            .try_collect::<Vec<CompletedPart>>()
```
[the remark this answers is not in the document]
Comment (oliverdaff) [reply src/s3/multipartcopy.rs]: I prefer the explicit types to help reduce the cognitive load on the reader of the code. However, we can remove one of the types and still achieve clarity:

```rust
let completed_parts: Vec<CompletedPart> = stream::iter(parts)
    .buffer_unordered(self.max_concurrent_uploads)
    .try_collect()
```

File: src/s3/s3_object.rs (hunk 4 above)
Comment (oliverdaff) [reply src/s3/s3_object.rs]: Exposing values through getters rather than directly accessing public fields ensures better encapsulation. However, removing the `pub` fields would break backward compatibility, so we must retain them.

**Comment by oliverdaff** [comment]
Review comments addressed.  This includes adding property tests to explore edge cases.

File: src/s3/multipartcopy.rs (hunk 1 above)
Comment (oliverdaff) [reply src/s3/multipartcopy.rs]: Refactored this code and exercised this with property tests and type checked values.

File: src/s3/multipartcopy.rs [hunk 6]
```diff
@@ -0,0 +1,930 @@
...
+        tracing::info!("Starting multipart copy");
+        let source_size =
+            get_source_size(&self.client, &self.source.bucket, &self.source.key).await?;
+
+        tracing::info!(
+            source_size = source_size.as_ref(),
+            part_size = self.part_size.as_ref(),
+        );
+
+        if source_size.as_ref() <= self.part_size.as_ref() {
```
Comment (timleslie) [inline src/s3/multipartcopy.rs:466]: **observation:** There's a slight contraction here with line `488`. We use `source <= part` here, but `source < part` (but spelt backwards) at `488` 🤔 I don't know if it makes much difference which one we change. If the block size is **equal to** the part size, then I guess an atomic copy will be more efficient? In that case we should change `488` to be `source <= part`. I'd also flip line `488` so that the inequality is spelt in the same order as `466`, just to make it easier to see how these two lines relate to each other.

**Review by timleslie — APPROVED** [review]
Looks good, one non-blocking observation that we might want to clean up, but other than that this is good to go 👍 The second pass clean up with additional tests really took this code up a notch, great work 🚀
